### ISSUE TITLE
perf(worker): 398 threads and 52% of worker CPU are libgomp teams parallelising a batch-1 LSTM (no OMP_NUM_THREADS clamp)

### ISSUE THREAD

**Issue by GoBeromsu** [SeniorAILab/SeeON-edge#449]
## Measured (live, `seeon-edge-final:a681ddc-epochfix`, 13 cameras, 14.2 h uptime, read-only probes)

```
python -m worker      : 398 threads   (py-spy: only 29 have a Python stack; 369 are native)
native child          : 270 threads   (exactly 20 per camera + 10 base)
process CPU           : 1.44 cores of 24; 55,944 CPU-s over 50,930 s wall
threads with 0 CPU-s in 14 h : 75
last thread created   : 13.9 h ago  -> static cost, NOT a leak (resampled 1/min for 10 min: flat 398)
```

## Accounting — 398 = 86 base + 13 cameras x 24

| Group | Count | Created at | Site |
| --- | --- | --- | --- |
| MainThread | 1 | 02:48:30 | `worker/__main__.py:508` |
| torch import pool | 23 | 02:48:31 | `import torch` (`worker/adapters/model/yolo_api.py:20`) |
| MainThread OpenMP team | 23 | 02:48:31 | libgomp, first CPU tensor op (`worker/adapters/model/warmup.py:60`) |
| OpenMP team #2 | 23 | 02:48:35 | libgomp (clip-recorder or child-monitor thread) |
| CUDA driver helper | 1 | 02:48:33 | `libcuda.so` |
| single long-lived workers (clip recorder, metadata/AU/preview/failure receivers, heartbeat, evidence sender, trace writer, status sender, MJPEG accept loop, restart watch, 2 fatal-event waiters, child monitor) | 15 | 02:48:35-39 | see trace |
| `worker-ingest-<camera>` | 13 | 02:48:39 | `worker/pipeline/ingest/lifecycle.py:465` |
| **per-camera OpenMP teams** | **299** | 02:48:42 - 03:05:25 | **libgomp, first CPU torch op in each ingest thread -> `worker/adapters/model/torch_lstm_fall.py:172-176`** |

The 299 arrive as **13 discrete blocks of exactly 23**, each at the moment that camera's RTSP stream delivered its first frame.

## Mechanism (confirmed by controlled experiment in the same image)

`nproc = 24`, `23 = nproc - 1`. libgomp creates a **per-calling-thread** team of `nproc-1` workers on that thread's first parallel region and keeps it for the thread's lifetime:

```
threads after start:                 1
threads after import torch:         24   (+23)
threads after 1st MainThread matmul: 47  (+23)
worker thread 1: 47 -> 71  (delta 24)   # 1 thread + 23 gomp workers
worker thread 2: 71 -> 95  (delta 24)
```

Trigger: `worker/adapters/model/torch_lstm_fall.py:92` defaults to `device: str = "cpu"` and runs a **batch-1** LSTM under `no_grad` at `:172`. There is **no `OMP_NUM_THREADS`, `MKL_NUM_THREADS`, `torch.set_num_threads`, or `cv2.setNumThreads` anywhere** in the repo, Dockerfile, or compose (grep is empty).

**52 % of all process CPU (29,015 of 55,944 CPU-s) is spent inside those OpenMP workers**, parallelising a batch-1 LSTM across 24 threads — work that cannot benefit from it; the barrier overhead exceeds the compute.

## Fix (recommended, not applied)

Set `OMP_NUM_THREADS=1` in the worker's compose environment (and/or `torch.set_num_threads(1)` in `worker/__main__.py` before the first model op; the env var is more reliable because it must land before the first parallel region).

Expected: **398 -> ~112 threads** (-286) and a direct cut into the 52 % of CPU above. Risk low: per-call CPU latency may rise slightly, aggregate throughput should improve.

## Not a leak, and the native side is proportional too

Native child: exactly **20 threads per camera** (1 `nvv4l2decoder`, 1 `task`, 2 `rtpjitterbuffer`, 2 `rtpsession-rtcp`, 2 `timer`, 2 `fakesrc:src`, 4 `udpsrc:src`, 6 `queue:src`) + 10 base = 270. `nvv4l2decoder` = exactly 13 and `task` = exactly 13, so no pipeline has leaked. GStreamer `GstTask` threads stop/restart on RTCP timers, which makes creation timestamps misleading — an earlier "leak" reading of that data was survivor bias.

Also: the TensorRT "workspace pool (4)" in `worker/native/deepstream/src/trt_perception.cpp:216-250` pools execution contexts and device buffers, **not threads** — it adds zero threads.

## Open question (the one thing this trace did not settle)

Whether this oversubscription causes the ~30 s half-speed windows every ~100 s observed during the #429 work. Evidence is mixed: total CPU is only 1.44/24 cores and a 214-sample 1 Hz series showed no ~100 s periodicity and no 2x dip; but if several cameras' inference windows phase-align, demand momentarily becomes 13 x 24 = 312 runnable threads on 24 cores, and per-camera OpenMP CPU is very uneven (199 s to 10,371 s).

Discriminating probe: sample `/proc/<pid>/task/*/stat` at 1 Hz for ~10 min, bucket by the creation-time cohort map (13 ingest vs 299 OpenMP), and plot against the worker's fps metric. Cheaper: at the next already-scheduled restart, bring the worker up with `OMP_NUM_THREADS=1` and see whether the windows disappear.

Related: #430 (single TensorRT execution context serialises inference), #428 (pacing ~11 fps), #448 (should the fall model move to the GPU).

**Comment by GoBeromsu** [comment]
## Host-only reproduction (no RTSP, no facility stack) — and a correction to the proposed fix

Reproduced the mechanism and measured both candidate fixes on a 24-core / RTX 5070 Ti host (`nproc = 24`, same as the facility reading, so the `23 = nproc-1` arithmetic reproduces exactly). Harness models the production call shape only: `(1, 30, 51)` → `nn.LSTM(51, 128, num_layers=2, batch_first)` → `Linear(128, 2)` → softmax under `no_grad`, geometry read from `models/fall/lstm/arch.json`. Random weights, shape-identical to `model.pt` — timing does not depend on weight values.

### 1. The fix works, and the win is larger than "a direct cut into 52 %"

13 threads × 100 inferences each, one shared module, as production does:

| | unclamped (today) | `OMP_NUM_THREADS=1` | delta |
| --- | --- | --- | --- |
| throughput | 1,490 inf/s | 14,120 inf/s | **9.5×** |
| ms per inference | 0.671 | 0.071 | **9.5×** |
| process CPU during the pass | 19.57 cores | 6.90 cores | **2.8× less** |
| p50 latency (isolated) | 0.364 ms | 0.276 ms | 1.3× |
| p99 latency (isolated) | **2.6 ms – 397 ms, run to run** | 0.285 ms | deterministic |

The tail is the part worth flagging. Unclamped, the same 300-sample pass produced p99 = 396.9 ms on one run and 2.57 ms on the next — the barrier storm's cost depends on who else is in a parallel region at that moment. Clamped, p99 = 0.285 ms with p95 = 0.282 ms; the distribution stops having a tail. For a fall-decision path, removing an unpredictable multi-hundred-ms stall is worth more than the throughput number.

### 2. Correction: `torch.set_num_threads(1)` in `worker/__main__.py` does **not** fix this

The issue offers the env var "and/or `torch.set_num_threads(1)` in `worker/__main__.py` before the first model op". The second option does not work, and the reason is structural: libgomp's `nthreads-var` is a **per-thread** ICV, and `torch.set_num_threads` calls `omp_set_num_threads`, which writes only the calling thread's copy. Measured — threads added by each of 4 worker threads at its first CPU tensor op:

| configuration | process threads at start | added per worker thread |
| --- | --- | --- |
| no clamp | 24 | **23, 17, 4, 13** |
| `torch.set_num_threads(1)` on MainThread only | 24 | **23, 1, 3, 3** |
| `torch.set_num_threads(1)` inside each worker thread | 24 | 0, 0, 0, 0 |
| `OMP_NUM_THREADS=1` in env | **1** | 0, 0, 0, 0 |

Two things follow:

- A MainThread-only clamp still lets `worker-ingest-<camera>` threads mint full teams — exactly the 299 threads this issue is about. To get the same result in Python you would have to call `torch.set_num_threads(1)` **inside every ingest thread**, i.e. in `worker/pipeline/ingest/lifecycle.py`'s thread body, not once at startup.
- The env var is strictly better than that: `start_threads = 1` vs `24`. It lands before `import torch`, so it also removes the **23-thread torch import pool** that the clamp-in-thread route leaves behind. Expected total is therefore below the issue's `~112` estimate.

Recommendation unchanged in direction, sharpened in form: set `OMP_NUM_THREADS=1` (and `MKL_NUM_THREADS=1`) in the worker's environment — image `ENV` and/or compose `environment` — and do not rely on a Python-side clamp. Grep confirms neither var, nor `torch.set_num_threads`, nor `cv2.setNumThreads` exists anywhere in `worker/`, `backend/`, `shared/`, `contracts/`, `scripts/`, `docs/`, `compose.edge*.yaml`, `Dockerfile.*`, or `.env.edge.prod.example`.

### 3. Caveats

- Host-only. The harness isolates the LSTM call; it does not reproduce the AU receive / remux / manifest / relay work that shares the real process, so absolute CPU figures are not the facility's.
- The GPU alternative was measured too and is worse here — see the comment on #448.
- The unclamped p99 range is genuinely unstable across runs; treat 397 ms as "this distribution has no bounded tail", not as a typical value.

**Comment by parkjunho12** [comment]
`jarvis:fix` 부착 — 24-core host에서 batch-1 LSTM 호출이 camera ingest thread마다 libgomp worker 23개를 생성하는 재현 경로와 OMP_NUM_THREADS=1 비교 측정(throughput 9.5배, thread team 제거)이 있어 동작/성능 결함 수정으로 분류합니다.
